### PR TITLE
Fix `backend.copyto` for mismatched dtypes to CuPy ndarray

### PR DESCRIPTION
**Pull request by hvy**
Fixes a parameter initialization bug. See https://github.com/chainer/chainer/pull/7965#issuecomment-525143855.

### CODE REVIEW

File: chainer/backend.py [hunk 1]
```diff
@@ -65,7 +65,8 @@ def copyto(dst, src):
     elif isinstance(dst, cuda.ndarray):
         if isinstance(src, chainer.get_cpu_array_types()):
             src = numpy.asarray(src)
-            if dst.flags.c_contiguous or dst.flags.f_contiguous:
+            if (src.dtype == dst.dtype
```
Comment (hvy) [inline chainer/backend.py]: `set` does not accept an array of different dtype, so we choose to fall through if dtypes don't match.

File: tests/chainer_tests/test_variable.py [hunk 2]
```diff
@@ -1837,11 +1837,13 @@ def setUp(self):
...
 
-    def check_constant_initialization(self, x, a, xp, expected_device):
+    def check_constant_initialization(
+            self, x, a, xp, expected_dtype, expected_device):
         x.initialize(a.shape)
         assert isinstance(x.data, xp.ndarray)
         assert x._has_chainerx_array is (xp is chainerx)
-        xp.testing.assert_array_equal(x.data, xp.asarray(a))
+        xp.testing.assert_array_equal(
+            x.data.astype(expected_dtype, copy=False), xp.asarray(a))
```
Comment (toslunar) [inline tests/chainer_tests/test_variable.py]: I suppose
```python
assert x.dtype == expected_dtype
xp.testing.assert_array_equal(x.data, xp.asarray(a, expected_dtype))
```

File: tests/chainer_tests/test_variable.py [hunk 3]
```diff
@@ -1837,11 +1837,13 @@ def setUp(self):
...
         xp.testing.assert_array_equal(x.grad, np.float32('nan'))
```
Comment (toslunar) [inline tests/chainer_tests/test_variable.py]: Why do you keep `np.float32`?  It could be `np.nan` or `float('nan')`.

**Comment by hvy** [comment]
Thanks for the comments. I'll address them in a separate PR. Removed `test_variable.py` from this one. PTAL.

**Comment by niboshi** [comment]
Should `copyto` support dtype casting?
I think `copyto` is a low-level function which usual users don't use.
I feel it's better to fix the caller to handle dtypes rather than adding convenient features into `copyto`.

**Comment by toslunar** [comment]
Random ideas:

- Automatic casting is error-prone, assuming that `backend.copyto` is used in model definitions more often than in training loops.
- `backend.copyto` could *always* check `dtype`s.
- `strides` (or just contiguity) might also matter.
- `numpy.copyto` supports casting.

**Comment by hvy** [comment]
I'm okay with making it strict, not allowing different dtypes. This functions is by no means a hot spot so we can afford some checking. We could also take inspiration from `numpy.copyto` and allow the caller to control the casting behavior, but that's a bit overkill considering the current use cases of this function.

@toslunar do you have concrete cases where stride checking would be beneficial?

**Comment by toslunar** [comment]
> @toslunar do you have concrete cases where stride checking would be beneficial?

Not really.  Loading NHWC-layout parameters, maybe.  A low-level `copy-to` sounds like `memcpy`.

**Comment by toslunar** [comment]
We discussed the status of the PR internally.
- The issue is classified as a bug because the behavior depends on the contiguity of the arrays.
- Since the `master` branch is stable now, it's better to align the behavior to the permissive side, i.e. allowing casting always (rather than denying it).

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by toslunar** [comment]
Jenkins, test this please.

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by toslunar** [comment]
FlexCI failures: timeout of `chainer/tests/chainer_tests/functions_tests/rnn_tests/test_function_slstm.py`

**Comment by emcastillo** [comment]
Re-running pfCI still fails on the same tests

[FAILED] /chainer/tests/chainer_tests/functions_tests/normalization_tests/test_l2_normalization.py (60 failed, 486 passed, 102 skipped, 648 deselected in 36.70 seconds)
--


Also now, both of them added some extra errors.
Will clone and verify locally 😢

**Comment by emcastillo** [comment]
Jenkins, test this please

**Comment by emcastillo** [comment]
Jenkins, test this please

**Pull request merged**